src/hotspot/share/runtime/basicLock.cpp

```diff
@@ -66,7 +66,7 @@ void BasicLock::move_to(oop obj, BasicLock* dest) {
   // is small (given the support for inflated fast-path locking in the fast_lock, etc)
   // we'll leave that optimization for another time.
 
-  if (displaced_header().is_neutral()) {
+  if (LockingMode == LM_LEGACY && displaced_header().is_neutral()) {
     // The object is locked and the resulting ObjectMonitor* will also be
```

Comment:
Isn't the set_displaced_header() below also unnecessary for other locking modes, i.e shouldn't it be `if (LockingMode != LM_LEGACY) { return; }`?

Reply:
Correct. We can skip copying the displaced header completely (also for OSR). I've updated the PR.